### DISCUSSION TITLE
Stability Issues / 5xx Errors

### DISCUSSION THREAD

**Discussion by AlekseyGorodetsky** [ngoduykhanh/PowerDNS-Admin#1736] [General]
Hello!

First of all, thank you for the necessary and important program - 111
The program really increases the productivity of the administrator, and therefore is important!

However, sometimes instability occurs in the operation of the program, which appears in the following pictures (attached to the message).
I also attach logs from the application container.

I'm using the containerner image:
https://hub.docker.com/r/powerdnsadmin/pda-legacy
I tried both tag options - latest, dev.

The logs are attached from a container with an image that has the tag - dev.

I tried for a long time to find the reason for the error - powerdnsadmin, powerdns or postgresql
But I assume that powerdns or postgresql is not a source of errors here.

Help me please...
![powerdnsadmin - Error 1](https://github.com/PowerDNS-Admin/PowerDNS-Admin/assets/16332401/d7dae0c0-945e-43a6-92cf-47967b78ebe7)
![powerdnsadmin - Error 2](https://github.com/PowerDNS-Admin/PowerDNS-Admin/assets/16332401/52bfc0f1-a285-4b05-a6f5-466470623232)
[powerdnsadmin.log](https://github.com/PowerDNS-Admin/PowerDNS-Admin/files/14038808/powerdnsadmin.log)

**Comment by AzorianMatt** [comment]
@AlekseyGorodetsky you said you had some error logs attached, but I don't see them.

**Reply by AlekseyGorodetsky** [reply]
This is the file attached to the message - powerdnsadmin.log

**Reply by AlekseyGorodetsky** [reply]
I can download this file again...
[powerdnsadmin.log](https://github.com/PowerDNS-Admin/PowerDNS-Admin/files/14044192/powerdnsadmin.log)

**Reply by AzorianMatt** [reply]
@AlekseyGorodetsky clear your sessions table and try to run it again before accessing the web interface. The exception shows pretty clearly to me that the built-in database upgrade process is failing due to a key constraint violation resulting from duplicate IDs in your sessions table.

**Comment by AlekseyGorodetsky** [comment]
I completely clear table "sessions" (delete all records in it), then reboot the container, but the error does not go away...
![powerdnsadmin - Error 3](https://github.com/PowerDNS-Admin/PowerDNS-Admin/assets/16332401/ad7d9449-a07e-402b-a823-501e81271935)
![powerdnsadmin - Error 4](https://github.com/PowerDNS-Admin/PowerDNS-Admin/assets/16332401/db211641-3478-4fce-890a-579d78ddda59)

**Reply by AzorianMatt** [reply]
@AlekseyGorodetsky well try posting more error logs back. That may or may not have fixed the first issue, but that in no way means there won't be more.

**Comment by AlekseyGorodetsky** [comment]
I am attaching new logs
[powerdnsadmin2.log](https://github.com/PowerDNS-Admin/PowerDNS-Admin/files/14045264/powerdnsadmin2.log)

**Comment by AlekseyGorodetsky** [comment]
![powerdnsadmin - DB](https://github.com/PowerDNS-Admin/PowerDNS-Admin/assets/16332401/72f7a296-7ead-4e7d-97eb-79e5a32ec8ba)

**Comment by AzorianMatt** [comment]
@AlekseyGorodetsky not sure why you are running off of the dev branch, but please switch to the production release. I'm not going to continue providing support on the code base that is unreleased.

To be perfectly fair, I don't really want to support this problem at all right now. Hopefully someone else comes along and wants to contribute. This edition of PDA has been frozen for good reason. It has a lot of problems 
, that have brought a lot of issues. I froze the project so that it can quit costing me precious time.

**Comment by AlekseyGorodetsky** [comment]
I understand you, and I don’t require an answer...
But it seems that this is not a problem with the latest releases.
I switched back to container tag latest, cleared the session table, rebooted the container, but the error remained.
Perhaps there is another reason...
It’s a shame, but your program is really good and it’s needed!

**Reply by AzorianMatt** [reply]
@AlekseyGorodetsky you may very well be right. Admittedly, we don't put a lot of focus into testing the Postgres support, and there have been some specific issues in the recent past related to Postgres deployments. I know we definitely ran into some unexpected problems when we initially made this change to database storage for sessions. Just another wonderful Flask plugim to disappoint me.

**Comment by cwchristerw** [comment]
I had similar issue and this was caused by CSRF token being invalid and updating doesn't work correctly to MariaDB database. I have changed SESSION_TYPE env variable to filesystem then it worked properly. I'm not even sure that filesystem is valid option because there is no documented values for that only sqlalchemy.

**Comment by Spidermanium** [comment]
I also have session problem with MariaDB. when SESSION_TYPE = 'sqlalchemy'.
I try 0.4.1 version no problem. Run in docker no problem. Any help?

File "/opt/web/powerdns-admin/flask/lib64/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1035, in _emit_update_statements
    raise orm_exc.StaleDataError(
sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'sessions' expected to update 1 row(s); 0 were matched.
[2025-06-26 17:04:17,220] [_internal.py:187] INFO - 10.87.0.89 - - [26/Jun/2025 17:04:17] "GET /static/assets/css/style.css HTTP/1.1" 500 -
File "/opt/web/powerdns-admin/flask/lib64/python3.9/site-packages/MySQLdb/connections.py", line 257, in query
    _mysql.connection.query(self, query)
sqlalchemy.exc.IntegrityError: (MySQLdb.IntegrityError) (1062, "Duplicate entry 'session:e68d454c-d3ce-44a7-9e54-9abbf92b8d36' for key 'session_id'")
[SQL: INSERT INTO sessions (session_id, data, expiry) VALUES (%s, %s, %s)]
[parameters: ('session:e68d454c-d3ce-44a7-9e54-9abbf92b8d36', b'\x80\x04\x95d\x00\x00\x00\x00\x00\x00\x00}\x94(\x8c\n_permanent\x94\x88\x8c\x0b_csrf_token\x94\x8c@923243e92f5d11945d678a9f5b0a90921ea17e9151faf20bf2cb0563d8e70db0\x94u.', datetime.datetime(2025, 6, 26, 17, 9, 8, 529830, tzinfo=datetime.timezone.utc))]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
[2025-06-26 17:09:10,186] [flask_seasurf.py:344] WARNING - Forbidden (CSRF token missing or incorrect.): /admin/setting/basic/allow_user_create_domain/toggle
[2025-06-26 17:09:10,220] [_internal.py:187] INFO - 10.87.0.89 - - [26/Jun/2025 17:09:10] "POST /admin/setting/basic/allow_user_create_domain/toggle HTTP/1.1" 403 -
[2025-06-26 17:09:11,253] [flask_seasurf.py:344] WARNING - Forbidden (CSRF token missing or incorrect.): /admin/setting/basic/allow_user_create_domain/toggle
[2025-06-26 17:09:11,285] [_internal.py:187] INFO - 10.87.0.89 - - [26/Jun/2025 17:09:11] "POST /admin/setting/basic/allow_user_create_domain/toggle HTTP/1.1" 403 -

Warkaround is set SESSION_TYPE = 'filesystem', but it is not right.

**Reply by Spidermanium** [reply]
I fix problem. Issue was in def clean_up_expired_sessions session.py. I change time to utc and it works! Problem was it timezone. When session created it was immedeatly deleted as expires, couse time was in past.

**Reply by Sakrizz** [reply]
Thank you for the tip.

I had similar problem with MariaDB and the timezone was the problem here as well. I did not fix it from the session.py end. Instead of changing anything in the code, just changing system time, mariadb time and PowerDNS-Admin time also does the same trick.

I have this as systemd service - added TZ line there:
**/etc/systemd/system/powerdns-admin.service**
Environment="TZ=UTC"

MariaDB to UTC time:
**/etc/mysql/mariadb.conf.d/50-server.cnf**
[mysqld]
default-time-zone = '+00:00'

**and also the system itself to UTC time:**
timedatectl set-timezone UTC
